DESIGN.md

```diff
@@ -0,0 +1,127 @@
+# Design rationale for SCIP
+
+Sourcegraph supports viewing and navigation code.
+This is similar to an IDE, except for the actual "editor" part.
+
+The design of SCIP is based on this primary motivating use case,
+as well as for fixing the pain points we encountered with using LSIF.[^1]
+
+SCIP is meant to be _transmission_ format for sending
+data from some producers to some consumers
+-- it is not meant as a _storage_ format for querying.
+Ideally, producers should be able to directly output SCIP
+instead of going through an intermediate format
+and then converting to SCIP for transmission.
+
+[^1]: Sourcegraph historically supported LSIF uploads
+as well as maintained our own LSIF indexers,
+but we ran into issues of development velocity,
+debugging, as well as indexer performance bottlenecks.
+
+## Goals
+
+### Primary goals
+
+- Support code navigation at the fidelity of state-of-the-art IDEs.
+  - Why: We want people to have an excellent experience navigating
+    their code inside Sourcegraph itself.
+- Ease of writing indexers (i.e. producers).
+  - Adding cross-repo navigation support should be easy.
+    - Why: Scaling for Sourcegraph customers using code spread across many repos,
+      as well as supporting code navigation for package ecosystems.
+  - Adding file-level incrementality should be easy.
+    - Why: Scaling for large monorepos.
+  - Making the indexer parallel should be easy
+    - Why: Scaling for large monorepos.
+- Robustness against indexer bugs: Incorrect code nav data for a certain entity
+  should have a limited blast radius.
+- Ease of debugging.
+
+## Non-goals
+
+- Support use cases involving code modifications.
+  - Why: Sourcegraph's code search and navigation has historically
+    focused on read-only use cases. Adding support for code modifications
+    introduces more complexity.
+- Ease of writing consumers.
+  - Why: We expect the number of SCIP producers to be much higher
+    than the number of consumers,
+    so it makes sense to optimize for producers.
+- Be as compact as possible in uncompressed form.
+  - Why: Modern general-purpose compression formats like gzip
+    and zstd are already very good in terms of both compression
+    speed and compression ratio.
```

Comment:
Not quite sure how to phrase it, but I'd maybe add that the speed and ratio are especially good on data _like_ SCIP, because the "wasteful" repetition of string symbol names is particularly amenable to compression.

Reply:
Added a footnote for this.